src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -24,10 +24,298 @@ ucs_status_t uct_tcp_sockcm_ep_disconnect(uct_ep_h ep, unsigned flags)
     return UCS_ERR_NOT_IMPLEMENTED;
 }
 
+static void uct_tcp_sockcm_ep_init_comm_ctx(uct_tcp_sockcm_ep_t *cep)
+{
+    cep->comm_ctx.offset = 0;
+    cep->comm_ctx.length = 0;
+}
+
+static void uct_tcp_sockcm_ep_handle_disconnect(uct_tcp_sockcm_ep_t *cep,
+                                                ucs_status_t status)
+{
+    uct_cm_remote_data_t remote_data;
+
+    /* remote peer disconnected */
+    ucs_debug("ep %p (fd=%d): remote peer disconnected", cep, cep->fd);
+    uct_tcp_sockcm_ep_init_comm_ctx(cep);
+
+    if (cep->state & UCT_TCP_SOCKCM_EP_ON_SERVER) {
+        uct_cm_ep_server_connect_cb(&cep->super, status);
+    } else {
+        ucs_assert(cep->state & UCT_TCP_SOCKCM_EP_ON_CLIENT);
+        remote_data.field_mask = 0;
+        uct_cm_ep_client_connect_cb(&cep->super, &remote_data, status);
+    }
+
+    /* TODO handle disconnect if the ep already invoked the connect_cb */
+}
+
+static int uct_tcp_sockcm_ep_is_tx_rx_done(uct_tcp_sockcm_ep_t *cep)
+{
+    ucs_assert((cep->comm_ctx.length != 0));
+    return (cep->comm_ctx.offset == cep->comm_ctx.length);
+}
+
+ucs_status_t uct_tcp_sockcm_ep_progress_send(uct_tcp_sockcm_ep_t *cep,
+                                             int is_async)
+{
+    ucs_status_t status;
+    size_t sent_length;
+
+    ucs_assert(cep->comm_ctx.offset < cep->comm_ctx.length);
+
+    sent_length = cep->comm_ctx.length - cep->comm_ctx.offset;
+
+    status = ucs_socket_send_nb(cep->fd,
```

Comment:
according to send() man page, seems send should set errno=ECONNRESET if the conenction was closed, and no need to call ucs_socket_is_connected() to check it